File: software/examples/delphi/ExampleThreshold.pas
program ExampleThreshold;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickletLaserRangeFinder;

type
  TExample = class
  private
    ipcon: TIPConnection;
    lrf: TBrickletLaserRangeFinder;
  public
    procedure ReachedCB(sender: TBrickletLaserRangeFinder; const distance: word);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'XYZ'; { Change to your UID }

var
  e: TExample;

{ Callback for distance greater than 20 cm }
procedure TExample.ReachedCB(sender: TBrickletLaserRangeFinder; const distance: word);
begin
  WriteLn(Format('Distance: %d cm.', [distance]));
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  lrf := TBrickletLaserRangeFinder.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Turn laser on and wait 250ms for very first measurement to be ready }
  lrf.EnableLaser;
  Sleep(250);

  { Get threshold callbacks with a debounce time of 10 seconds (10000ms) }
  lrf.SetDebouncePeriod(10000);

  { Register threshold reached callback to procedure ReachedCB }
  lrf.OnDistanceReached := {$ifdef FPC}@{$endif}ReachedCB;

  { Configure threshold for "greater than 20 cm" }
  lrf.SetDistanceCallbackThreshold('>', 20, 0);

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.

File: software/examples/ruby/example_threshold.rb
#!/usr/bin/env ruby
# -*- ruby encoding: utf-8 -*-

require 'tinkerforge/ip_connection'
require 'tinkerforge/bricklet_laser_range_finder'

include Tinkerforge

HOST = 'localhost'
PORT = 4223
UID = 'XYZ' # Change XYZ to the UID of your Laser Range Finder Bricklet

ipcon = IPConnection.new # Create IP connection
lrf = BrickletLaserRangeFinder.new UID, ipcon # Create device object

ipcon.connect HOST, PORT # Connect to brickd
# Don't use device before ipcon is connected

# Turn laser on and wait 250ms for very first measurement to be ready
lrf.enable_laser
sleep 0.25

# Get threshold callbacks with a debounce time of 10 seconds (10000ms)
lrf.set_debounce_period 10000

# Register distance reached callback (parameter has unit cm)
lrf.register_callback(BrickletLaserRangeFinder::CALLBACK_DISTANCE_REACHED) do |distance|
  puts "Distance: #{distance} cm"
end

# Configure threshold for distance "greater than 20 cm" (unit is cm)
lrf.set_distance_callback_threshold '>', 20, 0

puts 'Press key to exit'
$stdin.gets
lrf.disable_laser # Turn laser off
ipcon.disconnect
